### PR TITLE
feat: oauth intsropsection configurable timeout

### PR DESCRIPTION
**Pull request by pike1212**
## Related issue

#310 

## Proposed changes

Make http timeouts configurable in the oauth2 introspector

## Checklist

- [X] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [X] I have read the [security policy](../security/policy)
- [X] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

File: .schemas/config.schema.json [hunk 1]
```diff
@@ -678,6 +678,35 @@
               }
             }
           ]
+        },
+        "retry": {
+          "type": "object",
+          "properties": {
+            "conn_timeout": {
```
Comment (aeneasr) [inline .schemas/config.schema.json]: I like `connection_timeout` more but we are already using `max_delay` in another retry definition so I would like to use `max_delay` here as well: https://github.com/ory/oathkeeper/blob/2c310f95063233b2d48d105a476f674516336537/.schemas/config.schema.json#L827-L836

File: .schemas/config.schema.json [hunk 2]
```diff
@@ -678,6 +678,35 @@
...
+              "default": "500ms",
+              "pattern": "^[0-9]+(ns|us|ms|s|m|h)$",
+              "description": "Sets the connection timeout. If reached, the request will canceled be retried.",
+              "examples": [
+                "5s",
+                "5m",
+                "5h"
+              ]
+            },
+            "max_wait": {
```
Comment (aeneasr) [inline .schemas/config.schema.json]: Let's rename this to `give_up_after` for consistency.

File: .schemas/authenticators.oauth2_introspection.schema.json [hunk 3]
```diff
@@ -146,6 +146,35 @@
           }
         }
       ]
+    },
```
Comment (aeneasr) [inline .schemas/authenticators.oauth2_introspection.schema.json]: You can revert these changes as this file is no longer being used and is just kept for backwards compatibility

File: .schemas/config.schema.json [hunk 4]
```diff
@@ -678,6 +678,35 @@
               }
             }
           ]
+        },
+        "retry": {
```
Comment (aeneasr) [inline .schemas/config.schema.json]: We have another retry section here:

https://github.com/ory/oathkeeper/blob/2c310f95063233b2d48d105a476f674516336537/.schemas/config.schema.json#L823-L838

Let's make this a definition ([example](https://github.com/ory/oathkeeper/blob/2c310f95063233b2d48d105a476f674516336537/.schemas/config.schema.json#L7-L27)) instead and link it using `$ref` ([example](https://github.com/ory/oathkeeper/blob/2c310f95063233b2d48d105a476f674516336537/.schemas/config.schema.json#L203-L205))

File: .schemas/authenticators.oauth2_introspection.schema.json [hunk 5]
```diff
@@ -146,6 +146,35 @@
...
+    "retry": {
+      "type": "object",
+      "properties": {
```
Comment (aeneasr) [inline .schemas/authenticators.oauth2_introspection.schema.json]: Let's add:

```suggestion
      "additionalProperties": false,
      "properties": {
```

**Comment by hefekranz** [comment]
hmm, I was trying to use this feature and it seems like this only works if pre_authorization is enabled?
Is there a specific reason for that?

https://github.com/pike1212/oathkeeper/blob/519dc8d25c48ff0cd39dcd4e05e69616126e0a26/pipeline/authn/authenticator_oauth2_introspection.go#L169

**Comment by aeneasr** [comment]
Yup, that's not correct - would you be open to fix this? :)

**Comment by hefekranz** [comment]
sure, can do :)
